src/codemirror/extensions/autocomplete.ts

```diff
@@ -0,0 +1,180 @@
+import {
+  startCompletion,
+  CompletionContext,
+  CompletionResult,
+  Completion,
+  autocompletion,
+} from "@codemirror/autocomplete";
+import { extractEditorStateInfo } from "../utils/extensions-utils";
+import { AutocompletionsMetadata } from "@/types";
+
+function isNewTagContext(lineText: string): boolean {
+  return /<\w*$/.test(lineText);
+}
+
+function generateComponentNameCompletions(
+  partialName: string,
+  options: AutocompletionsMetadata
+): Completion[] {
+  return Object.keys(options)
+    .filter((componentName) =>
+      componentName.toLowerCase().startsWith(partialName.toLowerCase())
+    )
+    .map((componentName) => ({
+      label: componentName,
+      section: "Components",
+      type: "keyword",
+      apply: (view, completion, from, to) => {
+        // insert the component name and a space
+        const insertText = `${completion.label} `;
+        view.dispatch({
+          changes: { from, to, insert: insertText },
+          // move the cursor to the end of the new inserted text
+          selection: { anchor: from + insertText.length },
+        });
+
+        // we need to trigger autocompletion to immediately show prop completions after inserting the component name
+        requestAnimationFrame(() => {
+          startCompletion(view);
+        });
+      },
+    }));
+}
+
+function extractAlreadyUsedProps(
```

Comment:
"Extract" is misleading IMO, maybe "filter"?

Reply:
it is actually does extract, the filter is done on the result of the extraction
it is extracting used props from the string line